Accessible names for follow buttons

keshavbiswa opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
- This PR adds accessible names to follow buttons as part of enhancing follow button feature
- Hidden span tags are used to add accessible names such as, follow user, follow tag, etc.
## Related Tickets & Documents
- This PR addresses the first part of #14356 
- 
## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [x] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/WU0J9RYTNjKa6XrqB3/giphy.gif)

djuber commented:
CI failed with these two cases - I was able to recreate the failure locally. 

  1) OrganizationTag when given valid id_code renders the proper user name
     Failure/Error: expect(liquid.render).to include(organization.slug)
       expected "Liquid error: internal" to include "org120"
     # ./spec/liquid_tags/organization_tag_spec.rb:15:in `block (3 levels) in <main>'

  2) OrganizationTag when given valid id_code renders image html
     Failure/Error: expect(liquid.render).to include("<img")
       expected "Liquid error: internal" to include "<img"
     # ./spec/liquid_tags/organization_tag_spec.rb:20:in `block (3 levels) in <main>'

I'm looking further into this

djuber commented on app/views/organizations/_liquid.html.erb:
```diff
@@ -12,7 +12,11 @@
+    <h2>
+      <a href="/<%= organization.slug %>" class="ltag__user__link"><%= organization.name %></a>
+      <%= follow_button %>
+      <span class="hidden screen-reader-only">Follow organization: <%= organization.second %></span>
```
The issue with the failing tests is here - `organization.second` doesn't seem to be defined (so raises an error when called). 

What did you want the "Follow organization" span to read?

aitchiss reviewed (commented):
Thanks for taking this on! Unfortunately at the moment it's not working as expected, for a couple of reasons (not fully anticipated in the original issue 😄 ):

- The visually hidden `span`s aren't inside of the button content, so aren't influencing the the accessible name of the button. Also, the majority of the follow buttons in the app are created by the `follow_button` function in `application_helper.rb` so it's not super easy to inject the extra `span`
- Even if the visually hidden span _was_ inside the button element, the `followButtons.js` pack directly sets and would override the text content of the button as it optimistically changes e.g. Follow to Following

I know in the original issue I mentioned the visually hidden `span` being preferable for translation reasons but given the current setup, I think using aria-label would perhaps be more straightforward and robust.

If we use aria-label, we could update the `follow_button` helper I mentioned above to add the correct aria-label, and then in `followButtons.js` when we update the text content of the button, we can also update the aria-label attribute.

What do you think? How would you feel about reworking it like this?

keshavbiswa commented:
> If we use aria-label, we could update the `follow_button` helper I mentioned above to add the correct aria-label, and then in `followButtons.js` when we update the text content of the button, we can also update the aria-label attribute.

@aitchiss Yep, I think aria-lable makes better sense in this case, I can rework and make the changes no problem 👍🏻

keshavbiswa commented:
@aitchiss Addressed the changes, please review 🙇🏻

keshavbiswa commented:
Tests are passing locally. 🤷🏻‍♂️

aitchiss commented on app/helpers/application_helper.rb:
```diff
@@ -119,7 +119,7 @@ def sanitized_sidebar(text)
   end
 
-  def follow_button(followable, style = "full", classes = "")
+  def follow_button(followable, style = "full", classes = "", aria_label: "")
```
refactor (blocking) - I think it would be even better if we didn't have to worry about passing in the aria-label, given it is always going to take the format of `"Follow #{type_of_the_followable}: #{followable_name}"`

Can we update this helper function to dynamically create the correct aria-label based on the type and name of the followable?

keshavbiswa replied:
Why didn't I think of it 🤦🏻

aitchiss commented:
> @aitchiss So with #14356, my plan is to divide it into multiple small PRs, this one is only for the initial aria-label to be added to the elements. The rest of it, maybe cover it in separate tiny PRs?

> Let me know your thoughts on it.

I think that makes perfect sense @keshavbiswa - for this one then we'll need to go the route of updating the aria-label in the `followButtons.js` pack, so that we don't break any accessibility while we work on the toggle button part later 👍

aitchiss commented on app/helpers/application_helper.rb:
```diff
@@ -131,12 +133,12 @@ def follow_button(followable, style = "full", classes = "", aria_label: "")
         }
       },
       class: "crayons-btn follow-action-button whitespace-nowrap #{classes} #{user_follow}",
-      aria: { label: aria_label },
```
I think it would make sense to continue setting the `aria-label` here too, e.g.

```
aria: { label: "Follow #{followable_type}: #{followable_name}" },
```

Then the aria-label is available on initial page load, even before any of the JS is run.

aitchiss commented on app/javascript/packs/followButtons.js:
```diff
@@ -21,6 +21,20 @@ function addButtonFollowText(button, style) {
+ * @param {string} followType The followableType of the button.
+ * @param {string} followName The name of the followable to be followed.
+ */
+function addAriaLabelToButton(button, followType, followName) {
```
This is great, but at the moment this is only being set on the initial page load, and we will need to update the aria-label when the user clicks on it too, e.g. 

```Follow ${followType.toLowerCase()}: ${followName}```

will need to become

```Unfollow ${followType.toLowerCase()}: ${followName}```

when the button is updated to the "Following state". Similarly the frontend has some code to change a User follow button text to "Follow back" when appropriate, and we would need to then also update the `aria-label` to `Follow ${followType.toLowerCase()} back: ${followName}`

Check out `addButtonFollowText` and `addButtonFollowingText` for where this happens currently

aitchiss reviewed (commented):
This helper function to set the aria-label looks good - we just need to make sure the label is updated when the button text changes e.g. from Follow -> Following or Follow -> Follow back, and also update the Cypress tests with the new button name 👍

msarit commented:
Hello @keshavbiswa 👋🏾  Thank you for your work on this PR! 😄 
Are you still addressing the suggestions that @aitchiss made?

keshavbiswa commented:
> Hello @keshavbiswa 👋🏾  Thank you for your work on this PR! 😄 
> Are you still addressing the suggestions that @aitchiss made?

Hi, yes I'm still working on it 😊

keshavbiswa commented:
Need some help with Cypress 🙏🏻

keshavbiswa commented:
Also changing the aria label of button to `follow back` will be done in later PR, when I'll change the button title to `Follow Back`

aitchiss commented on app/javascript/packs/followButtons.js:
```diff
@@ -27,12 +32,24 @@ function addButtonFollowText(button, style) {
-    'aria-label',
-    `Follow ${followType.toLowerCase()}: ${followName}`,
-  );
+function addAriaLabelToButton(button, followType, followName, style = '') {
```
refactor (blocking) - it's definitely great having this as a helper function - but I can see we're actually calling this in a few places with the arguments the wrong way around (e.g. on line 24 we pass in followName and followType in the wrong order).

I think it would be easier to read if we pass these options in an object e.g.

```js
function addAriaLabelToButton({ button, followType, followName, style = '' }) {
```

and then when we call it we can be really explicit:

```js
addAriaLabelToButton({
  button, 
  followName: name, 
  followType: className, 
  style: 'follow'
});
```

aitchiss reviewed (commented):
This is looking good, and the change to the aria-label seems to be working well! I've noticed some small issues we'll need to sweep up though -

In the logged out state, when I land on the home page I'm seeing a lot of errors in the console relating to the aria-label helper method:

!["cannot read property 'toLowerCase' of undefined"](https://user-images.githubusercontent.com/20773163/130759109-c9b89bac-9c38-48b7-ba0d-9738fd8df7af.png)

I also noticed on a user's profile page, the Follow button aria-label is getting "undefined" as the username:

![screenshot of button location on user profile page](https://user-images.githubusercontent.com/20773163/130759277-b0b2b924-b954-41b7-8a61-87d32c158d09.png)

For the Cypress tests, I think you can actually update existing tests, rather than adding new ones. We have a lot of tests for the follow button functionality, but they all target the button by e.g. `findByRole('button', { name: 'Follow' })` - findByRole targets elements by accessible name, so wherever this appears, it will now need changed to the new aria-label.

If you search the codebase for `cy.findByRole('button', { name: 'Follow' })` you should find all the test files that need tweaked 🙂  Let me know if you have any difficulties!

citizen428 commented on app/helpers/application_helper.rb:
```diff
@@ -123,6 +123,8 @@ def follow_button(followable, style = "full", classes = "")
     return if followable == DELETED_USER
 
     user_follow = followable.instance_of?(User) ? "follow-user" : ""
+    followable_type = followable.decorated? ? followable.object.class.name.downcase : followable.class.name.downcase
```
Nitpick: We can save some repetition here:

```suggestion
    (followable.decorated? ? followable.object : followable).class.name.downcase
```

keshavbiswa commented:
@aitchiss All tests passing! Ready for review again.

aitchiss commented on cypress/integration/seededFlows/notificationsFlows/followUserFromNotifications.spec.js:
```diff
@@ -10,16 +10,16 @@ describe('Follow user from notifications', () => {
   it('Follows and unfollows a user from a follow notification', () => {
     cy.findByRole('heading', { name: 'Notifications' });
-    cy.intercept('/follows').as('followsRequest');
+    // cy.intercept('/follows').as('followsRequest');
```
I can't recall why we had this intercept here, but if we don't need it any more, go ahead and delete 🙂 
```suggestion
```

aitchiss commented on cypress/integration/seededFlows/tagsFlows/followTag.spec.js:
```diff
@@ -14,7 +14,7 @@ describe('Follow tag', () => {
     it('Follows and unfollows a tag from the tag index page', () => {
       cy.intercept('/follows').as('followsRequest');
-      cy.findAllByRole('button', { name: 'Follow' }).first().as('followButton');
+      cy.findByRole('button', { name: 'Follow tag: tag1' }).as('followButton');
```
I love that the unique names mean we no longer need this slightly hacky `first()` call 🔥

keshavbiswa replied:
> Nice work - thanks for fixing those tests! ✨ Excited to see this change in the codebase 😄

Super excited to get this merged 😄  Took more than a month to complete.

aitchiss reviewed (approved):
Nice work - thanks for fixing those tests! ✨  Excited to see this change in the codebase 😄
